tests/frontend/test_circuitspecs_x.py

```diff
@@ -0,0 +1,481 @@
+
+        res = prog.compile("X")
+        expected = expected.compile("X")
+        assert program_equivalence(res, expected, atol=tol)
+
+    def test_incorrect_s2gate_modes(self):
+        """Test exceptions raised if S2gates do not appear on correct modes"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+        n_modes = 8
+        half_n_modes = n_modes // 2
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE) | (q[0], q[1])
+            ops.S2gate(SQ_AMPLITUDE) | (q[2], q[3])
+            ops.S2gate(SQ_AMPLITUDE) | (q[4], q[5])
+            ops.S2gate(SQ_AMPLITUDE) | (q[7], q[6])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="The applied unitary cannot mix between the modes {}-{} and modes {}-{}.".format(
+                    0, half_n_modes - 1, half_n_modes, n_modes - 1
+                )):
+            res = prog.compile("X")
+
+    def test_incorrect_s2gate_params(self):
+        """Test exceptions raised if S2gates have illegal parameters"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+        wrong_sq_value = [np.round(SQ_AMPLITUDE+0.1, 4)]
+        allowed_sq_ranges = Ranges([0], [1.0], variable_name="r")
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE) | (q[0], q[4])
+            ops.S2gate(0) | (q[1], q[5])
+            ops.S2gate(SQ_AMPLITUDE) | (q[2], q[6])
+            ops.S2gate(SQ_AMPLITUDE+0.1) | (q[3], q[7])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match=r"Incorrect squeezing val"):
+            res = prog.compile("X")
+
+    def test_s2gate_repeated_modes(self):
+        """Test exceptions raised if S2gates are repeated"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE) | (q[0], q[4])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match=r"Incorrect squeezing val"):
+            res = prog.compile("X")
+
+    def test_s2gate_repeated_modes_half_squeezing(self):
+        """Test exceptions raised if S2gates are repeated"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE/2) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE/2) | (q[0], q[4])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        res = prog.compile("X")
```

Comment:
This test doesn't seem to be testing anything :/ Should the S2gates not be repeated, and instead it's just testing that half squeezing is not allowed?

Reply:
Quite the opposite, this tests shows the utility of the compiler, in particular by knowing that those two Squeezing gates by half the allowed amplitude can compiled into a single gate with parameter equal to allowed amplitude.

Reply:
Oh that's really cool! You could add in an assert to show that the compiled program has squeezing of 1.0